File: 3397.maximum-number-of-distinct-elements-after-operations/solution.kt
class Solution {
    fun maxDistinctElements(nums: IntArray, k: Int): Int {
        var p = Int.MIN_VALUE
        var res = 0

        for (num in nums.sorted()) {
            val q = maxOf(p, num - k)
            if (q <= num + k) {
                res++
                p = q + 1
            }
        }
        return res
    }
}

File: 3397.maximum-number-of-distinct-elements-after-operations/solution.go
func maxDistinctElements(nums []int, k int) int {
    res := 0
    p := math.MinInt
    slices.Sort(nums)
    for _, num := range nums {
        if q := max(p, num - k); q <= num + k {
            p = q + 1
            res++
        }
    }
    return res
}